### ISSUE TITLE
Spdylog: Crashes on Windows when window unloads

### ISSUE THREAD

**Issue by bpasero** [Microsoft/vscode#39659]
Steps to Reproduce:

1. open a folder
2. open another folder

=> you see a crash dialog popping up
=> it seems to be a race condition because it does not always reproduce

On Windows if you configure `set ELECTRON_DEFAULT_ERROR_MODE=1` you can manage to debug the crash in VS, but it needs the symbols.

The stack points to spdylog:

![image](https://user-images.githubusercontent.com/900690/33604277-99e264c4-d9b5-11e7-954a-7646ec61c1d1.png)

One theory is that the node.js environment goes down during window unload and spdylog raises an exception when this happens making the entire reload fail.

**Comment by sandy081** [comment]
It is happening always when I reload the current window

**Comment by sandy081** [comment]
First we added following fixes
- Convert the exception in the spdlog library from c++ to javascript
- Dispose the log service properly. While disposing flush and drop logs.

This fixed the crash, but window is still not loading on refresh because spdlog library was still holding on to the log file and throwing JS exception.

Hence added another fix to fallback to Noop log service to unblock from testing.

Will take decision tomorrow whether to disable logging or go for complete fix.

**Comment by joaomoreno** [comment]
@alexandrudima Thanks for @sandy081's mitigation, this is not a crash anymore. The worse case scenario that currently happens is a window reload loses logs, only on Windows. Removing the severity.

**Comment by joaomoreno** [comment]
Fun bug, in two parts.

First part, @sandy081 already covered: we weren't disposing the logger on reload, thus the file handle was kept alive across reloads. Windows fails to open the file at that point, since there are open handles to it already.

Second part, we noticed that the bug kept happening even though we were properly disposing the logger, which closed the file descriptor, yet somehow the following `open` call would still fail. Using Process Hacker, we found out that log file handles are somehow inherited by the renderer's child processes (extension host, crash reporter). What is so problematic about this is that on reload, the crash reporter stays alive and goes orphan but keeps a handle to the log file. When the new renderer comes up, it can't open that same file because of the crash reporter process keeping that handle open. This explains why it doesn't reproduce in development: there's no crash reporter in development.

We figured out how to fix this deep in spdlog and created a PR: https://github.com/gabime/spdlog/pull/570

We already adopted that same fix on our node bindings to the library on `master`.
